app/src/androidTest/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TaskDaoTest.kt

```diff
@@ -42,10 +40,6 @@ class TaskDaoTest {
         ToDoDatabase::class.java
     ).allowMainThreadQueries().build()
 
-    // Set the main coroutines dispatcher for unit testing.
-    @get:Rule
-    val mainCoroutineRule = MainCoroutineRule()
-
     // Ensure that we use an empty database for each test.
     @Before
     fun initDb() = database.clearAllTables()
```

Comment:
This shouldn't be needed either, a new database will be created per test

Reply:
What magic causes the database to be re-created for each test? It's only defined once in the test class.

Reply:
It might be the same instance in all tests but the `inMemoryDatabase` is cleared when the process is killed. If that happens after every test run, that's why we might see an empty database.

IMO, I think it's better to avoid these variables in the test class itself. The initialization should be done in the `@Before` methods. I'd have something like this:

```
class TaskDaoTest {
    lateinit var database: ToDoDatabase

    @Before
    fun init() {
        database = Room.inMemoryDatabase(...)
    }
}
```

My understanding is that if the same instance is reused across tests, that wouldn't make them elegible to be run in parallel. 

Maybe @JoseAlcerreca can chime in here as well

Reply:
Code updated to use your approach - it's safer than just clearing the tables.